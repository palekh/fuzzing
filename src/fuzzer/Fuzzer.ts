import {Collector, ICollector} from "../collector/Collector";
import {IResultTyped} from "../types/result.type";

export interface IFuzzer {
    fuzz(): IFuzzer;

    successes(): IResultTyped[];

    errors(): IResultTyped[];

    warnings(): IResultTyped[];

    all(): IResultTyped[];
}

export type IFuzzerParams = any[];

export abstract class Fuzzer implements IFuzzer {
    protected readonly params: IFuzzerParams = [];
    protected collector: ICollector = Collector.create();

    public abstract fuzz(): Fuzzer;

    public successes(): IResultTyped[] {
        return this.collector.getSuccesses();
    }

    public errors(): IResultTyped[] {
        return this.collector.getErrors();
    }

    public warnings(): IResultTyped[] {
        return this.collector.getWarnings();
    }

    public all(): IResultTyped[] {
        return this.collector.getAll();
    }

    protected abstract fuzzIteration(args: any): void;
}
