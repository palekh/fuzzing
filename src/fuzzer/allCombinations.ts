
export function allCombinations(input: any[][]): any[][] {
    const counters = input.map(() => 0);

    const output: any[][] = [];

    while (counters.length === input.length) {
        const iteration: any = [];
        counters.forEach((counter, index) => {
            iteration.push(input[index][counter]);
        });
        output.push(iteration);

        counters[0] += 1;
        counters.forEach((counter, index) => {
            const max = input[index].length;
            if (counter >= max) {
                counters[index] = 0;
                counters[index + 1] += 1;
            }
        });
    }

    return output;
}
