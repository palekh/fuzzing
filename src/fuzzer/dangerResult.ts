export function getDangerResults(): any {
    return [
        Number.NaN,
        undefined,
        null,
    ];
}

export function isDangerResult(result: any): boolean {
    const results = getDangerResults();

    if (isNaN(result)) {
        return true;
    }

    for (let i = 1; i < results.length; i += 1) {
        if (results[i] === result) {
            return true;
        }
    }

    return false;
}
