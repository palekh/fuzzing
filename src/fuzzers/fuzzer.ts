import {AnyFunction} from "../_types/anyFunction.type";

export interface IFuzzer {
    fuzz(): IFuzzer;
    getErrors(): Error[];
    getWarnings(): Error[];
}

export abstract class Fuzzer implements IFuzzer {
    protected readonly testSet: AnyFunction[];
    protected errors: Error[] = [];
    protected warnings: Error[] = [];

    protected constructor(...args: any[]) {
        // empty
    }

    public abstract fuzz(): Fuzzer;

    public getErrors(): Error[] {
        return this.errors;
    }

    public getWarnings(): Error[] {
        return this.warnings;
    }
}
