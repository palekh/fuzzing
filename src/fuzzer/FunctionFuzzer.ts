import {getNull} from "../checker/null";
import {getUndefined} from "../checker/undefined";
import {isDangerResult} from "../fuzzer/dangerResult";
import {Fuzzer, IFuzzer, IFuzzerParams} from "../fuzzer/Fuzzer";
import {AnyFunction} from "../types/anyFunction.type";

export class FunctionFuzzer extends Fuzzer implements IFuzzer {
    protected readonly params: IFuzzerParams;
    private readonly func: AnyFunction;

    private constructor(func: AnyFunction, params: IFuzzerParams) {
        super();
        this.func = func;
        this.params = params;
    }

    public static create(func: AnyFunction, params: IFuzzerParams): FunctionFuzzer {
        return new FunctionFuzzer(func, [
            getUndefined(),
            getNull(),
            ...params,
        ]);
    }

    public fuzz(): this {
        for (let i = 0; i < this.params.length; i += 1) {
            const parameter = this.params[i];
            this.fuzzIteration(parameter);
        }

        return this;
    }

    protected fuzzIteration(input: any): void {
        try {
            const result = this.func(input);

            if (isDangerResult(result)) {
                this.collector.addWarning({
                    description: "WARNING: Function returned result might be nullable or dangerous in some way",
                    input,
                    result,
                });
            } else {
                this.collector.addSuccess({
                    description: "SUCCESS: Function returned result is OK and no errors happened",
                    input,
                    result,
                });
            }
        } catch (error) {
            this.collector.addError({
                description: "FAILED: Function execution failed, check error stack trace",
                error,
                input,
            });
        }
    }
}
