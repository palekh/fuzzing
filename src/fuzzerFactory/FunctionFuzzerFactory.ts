import {cornerCaseBooleanArray, nonConsequentBooleanArray} from "../checker/boolean";
import {cornerCaseNumberArray, mathNumberArray, nonConsequentNumberArray} from "../checker/number";
import {cornerCaseObjectArray, nonConsequentObjectArray} from "../checker/object";
import {cornerCaseStringArray, nonConsequentStringArray} from "../checker/string";
import {FunctionFuzzer} from "../fuzzer/FunctionFuzzer";
import {IFuzzer, IFuzzerParams} from "../fuzzer/Fuzzer";
import {AnyFunction} from "../types/anyFunction.type";

export interface IFunctionFuzzerFactory {
    number(): IFuzzer;

    numberArray(): IFuzzer;

    string(): IFuzzer;

    stringArray(): IFuzzer;

    all(): IFuzzer;
}

export class FunctionFuzzerFactory implements IFunctionFuzzerFactory {
    private readonly func: AnyFunction;

    private constructor(func: AnyFunction) {
        this.func = func;
    }

    public static create(func: AnyFunction): FunctionFuzzerFactory {
        return new FunctionFuzzerFactory(func);
    }

    public boolean(): IFuzzer {
        return this.createFuzzer([
            ...cornerCaseBooleanArray(),
        ]);
    }

    public booleanArray(): IFuzzer {
        return this.createFuzzer([
            cornerCaseBooleanArray(),
            nonConsequentBooleanArray(),
        ]);
    }

    public number(): IFuzzer {
        return this.createFuzzer([
            ...cornerCaseNumberArray(),
            ...mathNumberArray(),
        ]);
    }

    public numberArray(): IFuzzer {
        return this.createFuzzer([
            cornerCaseNumberArray(),
            mathNumberArray(),
            nonConsequentNumberArray(),
        ]);
    }

    public string(): IFuzzer {
        return this.createFuzzer([
            ...cornerCaseStringArray(),
        ]);
    }

    public stringArray(): IFuzzer {
        return this.createFuzzer([
            cornerCaseStringArray(),
            nonConsequentStringArray(),
        ]);
    }

    public object(): IFuzzer {
        return this.createFuzzer([
            ...cornerCaseObjectArray(),
        ]);
    }

    public objectArray(): IFuzzer {
        return this.createFuzzer([
            cornerCaseObjectArray(),
            nonConsequentObjectArray(),
        ]);
    }

    public all(): IFuzzer {
        return this.createFuzzer([
            ...cornerCaseBooleanArray(),
            cornerCaseBooleanArray(),
            nonConsequentBooleanArray(),

            ...cornerCaseNumberArray(),
            ...mathNumberArray(),
            cornerCaseNumberArray(),
            mathNumberArray(),
            nonConsequentNumberArray(),

            ...cornerCaseStringArray(),
            cornerCaseStringArray(),
            nonConsequentStringArray(),

            ...cornerCaseObjectArray(),
            cornerCaseObjectArray(),
            nonConsequentObjectArray(),
        ]);
    }

    private createFuzzer(input: IFuzzerParams): IFuzzer {
        return FunctionFuzzer
            .create(this.func, input)
            .fuzz();
    }
}
