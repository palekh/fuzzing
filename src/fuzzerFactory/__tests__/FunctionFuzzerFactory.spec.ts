import {FunctionFuzzer} from "../../fuzzer/FunctionFuzzer";
import {fuzz} from "../../index";
import {sum} from "../__mocks__/sum.mock";
import {FunctionFuzzerFactory} from "../FunctionFuzzerFactory";

describe("FunctionFuzzerFactory", () => {
    const sumFuzzerFactory = fuzz(sum);

    test("functionFuzzer should create function fuzzer instance", () => {
        expect(sumFuzzerFactory instanceof FunctionFuzzerFactory).toBe(true);
    });

    test("functionFuzzer should create fuzzer with internal func", () => {
        expect((sumFuzzerFactory as any).func).toBe(sum);
    });

    const methodNames: Array<keyof FunctionFuzzerFactory> = [
        "boolean", "booleanArray", "number", "numberArray", "string", "stringArray", "all",
    ];

    methodNames.forEach(method => {
        test(`factory method: ${method} should creates fuzzer without error`, () => {
            expect((sumFuzzerFactory as any)[method]() instanceof FunctionFuzzer).toBe(true);
            expect(() => (sumFuzzerFactory as any)[method]()).not.toThrow();
        });
    });

    test("should create snapshots with correct error descriptions", () => {
        const errors = fuzz(sum)
            .numberArray()
            .all();

        expect(errors).toMatchSnapshot();
    });
});
