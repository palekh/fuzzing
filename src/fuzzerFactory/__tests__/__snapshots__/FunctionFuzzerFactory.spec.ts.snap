// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FunctionFuzzerFactory should create snapshots with correct error descriptions 1`] = `
Array [
  Object {
    "description": "Normal result.",
    "input": "",
    "result": 0,
    "type": "success",
  },
  Object {
    "description": "Normal result.",
    "input": "0",
    "result": "00",
    "type": "success",
  },
  Object {
    "description": "Normal result.",
    "input": "1",
    "result": "01",
    "type": "success",
  },
  Object {
    "description": "Normal result.",
    "input": "1234567890",
    "result": "01234567890",
    "type": "success",
  },
  Object {
    "description": "Danger result: 0UPPER_CASE. Check function implementation",
    "input": "UPPER_CASE",
    "result": "0UPPER_CASE",
    "type": "warning",
  },
  Object {
    "description": "Danger result: 0lower_case. Check function implementation",
    "input": "lower_case",
    "result": "0lower_case",
    "type": "warning",
  },
  Object {
    "description": "Danger result: 0CamelCase. Check function implementation",
    "input": "CamelCase",
    "result": "0CamelCase",
    "type": "warning",
  },
  Object {
    "description": "Danger result: 0!@#$%^&*()_+-=\`~[]\\\\{}|;':,./<>?. Check function implementation",
    "input": "!@#$%^&*()_+-=\`~[]\\\\{}|;':,./<>?",
    "result": "0!@#$%^&*()_+-=\`~[]\\\\{}|;':,./<>?",
    "type": "warning",
  },
  Object {
    "description": "Danger result: 0aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa. Check function implementation",
    "input": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
    "result": "0aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
    "type": "warning",
  },
  Object {
    "description": "Function execution failed.",
    "error": [TypeError: Cannot read property 'length' of undefined],
    "input": undefined,
    "type": "error",
  },
  Object {
    "description": "Function execution failed.",
    "error": [TypeError: Cannot read property 'length' of null],
    "input": null,
    "type": "error",
  },
]
`;
